src/backend/distributed/planner/query_pushdown_planning.c

```diff
@@ -1583,13 +1586,28 @@ FlattenJoinVars(List *columnList, Query *queryTree)
 				root->hasJoinRTEs = true;
 			}
 
-			normalizedVar = (Var *) flatten_join_alias_vars(root, (Node *) column);
+
+			normalizedNode = flatten_join_alias_vars(root, (Node *) column);
 
 			/*
 			 * We need to copy values over existing one to make sure it is updated on
 			 * respective places.
 			 */
-			memcpy(column, normalizedVar, sizeof(Var));
+
+			if (IsA(normalizedNode, Var))
+			{
+				memcpy(column, normalizedNode, sizeof(Var));
+			}
+			else if (IsA(normalizedNode, CoalesceExpr))
+			{
+				column = repalloc(column, sizeof(CoalesceExpr));
```

Comment:
As discussed in slack; I think this need to be reworked where the MemoryContext is set to the context that column comes from before calling `flatten_join_alias_vars`.

That function goes into `flatten_join_alias_vars_mutator` which recurses through the object graph and makes a deep copy besides some mutations.

Since `CoalesceExpr` contains a `List *` if will hit:https://github.com/postgres/postgres/blob/97c39498e5ca9208d3de5a443a2282923619bf91/src/backend/nodes/nodeFuncs.c#L2906-L2924 which deep copies the list into the `CurrentMemoryContext`.

So even though we memcpy the contents of `CoalesceExpr` into a struct that lives in the original MemoryContext it points to a `List *` in the `CurrentMemoryContext`.

Reply:
This `repalloc` was outcome of an optimization/enhancement to the query plan. However, given that it complicates things a lot, I'd prefer not to have `repalloc` or anything equivalent at all, and keep `CoalesceExpr ` on the extentedOpNode as well.

So, in the end, with this repalloc, we're targeting to get something like:

```SQL
SELECT worker_column_1  as id FROM (SELECT Coalesce(table_1.id, table_2.id)) FROM table_1 FULL JOIN t2 USING(id))
```

Without re-palloc, we'd have a redundant `Coalesce` on the master query/top level worker query, which doesn't affect the correctness, but mostly better for readability/debuggibility:
```SQL
SELECT Coalesce(worker_column_1, worker_column_1)  as id FROM (SELECT Coalesce(table_1.id, table_2.id)) FROM table_1 FULL JOIN t2 USING(id))
```

So, @velioglu unless you find a simple implementation for re-allocating the `Coalesce` to `Var`, I'd rather not go that router.

Reply:
Could we build a new list and return it? I don't think the `repalloc` is correct since it may change the pointer and we're not doing anything with that pointer.

Reply:
> Could we build a new list and return it? 

I think no, because the items in the list are actually the `Var`s in the `targetList` and `havingClause` entries, which are used later on the extended op node as well. So, it seems safer to keep `Coalesce ` and get rid of `repalloc`, only use `memcpy` after creating the correct `Coalesce `.